### PR TITLE
[Feature] Add multi-grid training.

### PR DESCRIPTION
**Pull request by SuX97**
This PR continues from #29 .

+ Add precise bn

### CODE REVIEW

**Comment by donnyyou** [comment]
Sorry! Another important issue has been ignored. The slowfast models of your is slightly different, compared to the models of PySlowFast. The number of subbns is 106 for yours, but 110 for PySlowFast. The inference time of PySlowFast models seems even faster, and the performance is about 1% higher with all the models with/without multi-grid training. @SuX97 @dreamerlin

**Comment by donnyyou** [comment]
I haven't test the inference speed, just for observation.

**Comment by SuX97** [comment]
> Sorry! Another important issue has been ignored. The slowfast models of your is slightly different, compared to the models of PySlowFast. The number of subbns is 106 for yours, but 110 for PySlowFast. The inference time of PySlowFast models seems even faster, and the performance is about 1% higher with all the models with/without multi-grid training. @SuX97 @dreamerlin

Thanks for reporting, @donnyyou ! I still have a confusion. In PySlowfast code, the precise bn use the same dataloader(a long-cycle dataloader same as the training one) but with ShortCycle Sampler closed:
```
    if cfg.MULTIGRID.SHORT_CYCLE and split in ["train"] and not is_precise_bn:
        # Create a sampler for multi-process training
        sampler = utils.create_sampler(dataset, shuffle, cfg)
        batch_sampler = ShortCycleBatchSampler(
            sampler, batch_size=batch_size, drop_last=drop_last, cfg=cfg
        )
        # Create a loader
        loader = torch.utils.data.DataLoader(
            dataset,
            batch_sampler=batch_sampler,
            num_workers=cfg.DATA_LOADER.NUM_WORKERS,
            pin_memory=cfg.DATA_LOADER.PIN_MEMORY,
            worker_init_fn=utils.loader_worker_init_fn(dataset),
        )
    else:
        # Create a sampler for multi-process training
        sampler = utils.create_sampler(dataset, shuffle, cfg)
        # Create a loader
        loader = torch.utils.data.DataLoader(
            dataset,
            batch_size=batch_size,
            shuffle=(False if sampler else shuffle),
            sampler=sampler,
            num_workers=cfg.DATA_LOADER.NUM_WORKERS,
            pin_memory=cfg.DATA_LOADER.PIN_MEMORY,
            drop_last=drop_last,
            collate_fn=detection_collate if cfg.DETECTION.ENABLE else None,
            worker_init_fn=utils.loader_worker_init_fn(dataset),
        )
    return loader
```
Did your implementation follows this setting? It seems strange as the dataloader of each round of precise bn in not identical to the training one.

**Comment by SuX97** [comment]
> Sorry! Another important issue has been ignored. The slowfast models of your is slightly different, compared to the models of PySlowFast. The number of subbns is 106 for yours, but 110 for PySlowFast. The inference time of PySlowFast models seems even faster, and the performance is about 1% higher with all the models with/without multi-grid training. @SuX97 @dreamerlin

Right, the difference lies in the lateral connection layer, as we do not use bn and relu after the conv layer:
```
                setattr(
                    self, lateral_name,
                    ConvModule(
                        self.inplanes // self.channel_ratio,
                        self.inplanes * 2 // self.channel_ratio,
                        kernel_size=(fusion_kernel, 1, 1),
                        stride=(self.speed_ratio, 1, 1),
                        padding=((fusion_kernel - 1) // 2, 0, 0),
                        bias=False,
                        conv_cfg=self.conv_cfg,
                        norm_cfg=None, # it is None here
                        act_cfg=None))
```
While in Pyslowfast:
```
class FuseFastToSlow(nn.Module):

...

        self.conv_f2s = nn.Conv3d(
            dim_in,
            dim_in * fusion_conv_channel_ratio,
            kernel_size=[fusion_kernel, 1, 1],
            stride=[alpha, 1, 1],
            padding=[fusion_kernel // 2, 0, 0],
            bias=False,
        )
        self.bn = norm_module(
            num_features=dim_in * fusion_conv_channel_ratio,
            eps=eps,
            momentum=bn_mmt,
        )
        self.relu = nn.ReLU(inplace_relu)
```

**Comment by donnyyou** [comment]
> > Sorry! Another important issue has been ignored. The slowfast models of your is slightly different, compared to the models of PySlowFast. The number of subbns is 106 for yours, but 110 for PySlowFast. The inference time of PySlowFast models seems even faster, and the performance is about 1% higher with all the models with/without multi-grid training. @SuX97 @dreamerlin
> 
> Right, the difference lies in the lateral connection layer, as we do not use bn and relu after the conv layer:
> 
> ```
>                 setattr(
>                     self, lateral_name,
>                     ConvModule(
>                         self.inplanes // self.channel_ratio,
>                         self.inplanes * 2 // self.channel_ratio,
>                         kernel_size=(fusion_kernel, 1, 1),
>                         stride=(self.speed_ratio, 1, 1),
>                         padding=((fusion_kernel - 1) // 2, 0, 0),
>                         bias=False,
>                         conv_cfg=self.conv_cfg,
>                         norm_cfg=None, # it is None here
>                         act_cfg=None))
> ```
> 
> While in Pyslowfast:
> 
> ```
> class FuseFastToSlow(nn.Module):
> 
> ...
> 
>         self.conv_f2s = nn.Conv3d(
>             dim_in,
>             dim_in * fusion_conv_channel_ratio,
>             kernel_size=[fusion_kernel, 1, 1],
>             stride=[alpha, 1, 1],
>             padding=[fusion_kernel // 2, 0, 0],
>             bias=False,
>         )
>         self.bn = norm_module(
>             num_features=dim_in * fusion_conv_channel_ratio,
>             eps=eps,
>             momentum=bn_mmt,
>         )
>         self.relu = nn.ReLU(inplace_relu)
> ```

This will cause some performance drop.

**Comment by donnyyou** [comment]
> > Sorry! Another important issue has been ignored. The slowfast models of your is slightly different, compared to the models of PySlowFast. The number of subbns is 106 for yours, but 110 for PySlowFast. The inference time of PySlowFast models seems even faster, and the performance is about 1% higher with all the models with/without multi-grid training. @SuX97 @dreamerlin
> 
> Thanks for reporting, @donnyyou ! I still have a confusion. In PySlowfast code, the precise bn use the same dataloader(a long-cycle dataloader same as the training one) but with ShortCycle Sampler closed:
> 
> ```
>     if cfg.MULTIGRID.SHORT_CYCLE and split in ["train"] and not is_precise_bn:
>         # Create a sampler for multi-process training
>         sampler = utils.create_sampler(dataset, shuffle, cfg)
>         batch_sampler = ShortCycleBatchSampler(
>             sampler, batch_size=batch_size, drop_last=drop_last, cfg=cfg
>         )
>         # Create a loader
>         loader = torch.utils.data.DataLoader(
>             dataset,
>             batch_sampler=batch_sampler,
>             num_workers=cfg.DATA_LOADER.NUM_WORKERS,
>             pin_memory=cfg.DATA_LOADER.PIN_MEMORY,
>             worker_init_fn=utils.loader_worker_init_fn(dataset),
>         )
>     else:
>         # Create a sampler for multi-process training
>         sampler = utils.create_sampler(dataset, shuffle, cfg)
>         # Create a loader
>         loader = torch.utils.data.DataLoader(
>             dataset,
>             batch_size=batch_size,
>             shuffle=(False if sampler else shuffle),
>             sampler=sampler,
>             num_workers=cfg.DATA_LOADER.NUM_WORKERS,
>             pin_memory=cfg.DATA_LOADER.PIN_MEMORY,
>             drop_last=drop_last,
>             collate_fn=detection_collate if cfg.DETECTION.ENABLE else None,
>             worker_init_fn=utils.loader_worker_init_fn(dataset),
>         )
>     return loader
> ```
> 
> Did your implementation follow this setting? It seems strange as the dataloader of each round of precise bn is not identical to the training one.

I totally follow the setting of PySlowFast.

**Comment by SuX97** [comment]
PR closed. Failed to reproduce the accuracy, let's just try to merge precise bn as a harmless tool.
